geoengine/colorizer.py

```diff
@@ -64,44 +101,149 @@ def __init__(
-            n_steps (int): how many steps the colormap should have.
-
-        Returns:
-            str: Returns the geoengine compatible json configuration string.
-        """
         # get the map, and transform it to [0,255] values
-        colormap = ScalarMappable(cmap=self.__map_name).to_rgba(
-            np.linspace(self.__min_max[0], self.__min_max[1], self.__n_steps), bytes=True).tolist()
+        colormap = [
+            (int(x[0]), int(x[1]), int(x[2]), int(x[3])) for x in ScalarMappable(cmap=map_name).to_rgba(
+                np.linspace(min_max[0], min_max[1], n_steps), bytes=True)]
 
         # if you want to remap the colors, you can do it here (e.g. cutting of the most extreme colors)
         value_bounds = [
-            int(x) for x in np.linspace(self.__min_max[0], self.__min_max[1], self.__n_steps).tolist()
+            int(x) for x in np.linspace(min_max[0], min_max[1], n_steps)
         ]
 
         # generate color map steps for geoengine
         breakpoints = [
-            {"value": value, "color": color} for (value, color) in zip(value_bounds, colormap)
+            ColorBreakpoint(color=color, value=value) for (value, color) in zip(value_bounds, colormap)
         ]
 
-        # create the json string for geoengine
-        colorizer = "custom:" + json.dumps(
-            {
-                "type": "linearGradient",
-                "breakpoints": breakpoints,
-                "noDataColor": self.__no_data_color,
-                "defaultColor": self.__default_color,
-            }
+        colorizer = LinearGradientColorizer(
+            breakpoints=breakpoints,
+            no_data_color=no_data_color,
+            default_color=default_color
         )
 
         return colorizer
+
+    @ abstractmethod
+    def to_api_dict(self) -> api.Colorizer:
+        pass
+
+    def to_json(self) -> str:
+        """Return the colorizer as a JSON string."""
+        return json.dumps(self.to_api_dict())
+
+    @staticmethod
+    def from_response(response: api.Colorizer) -> "Colorizer":
+        """Create a colorizer from a response."""
+        if response['type'] == 'linearGradient':
+            return LinearGradientColorizer.from_response_linear(cast(api.LinearGradientColorizer, response))
+        if response['type'] == 'palette':
+            return PaletteColorizer.from_response_palette(cast(api.PaletteColorizer, response))
+        if response['type'] == 'logarithmicGradient':
+            return LogarithmicGradientColorizer.from_response_logarithmic(
+                cast(api.LogarithmicGradientColorizer, response)
+            )
+
+        raise TypeError(f"Unknown colorizer type: {response['type']}")
+
+
+class LinearGradientColorizer(Colorizer):
+    '''A linear gradient colorizer.'''
+    breakpoints: List[ColorBreakpoint]
+
+    def __init__(
+        self,
+        breakpoints: List[ColorBreakpoint],
+        no_data_color: Tuple[int, int, int, int],
+        default_color: Tuple[int, int, int, int]
```

Comment:
Should we directly use over/under color here?

Reply:
lets do another PR